lib/tokenlist.cpp

```diff
@@ -485,6 +485,10 @@ static Token * skipDecl(Token *tok)
         } else if (Token::Match(vartok, "%var% [:=(]")) {
             return vartok;
         }
+        else if (Token::simpleMatch(vartok, "decltype ( auto )")) {
```

Comment:
I guess it's a good idea to tweak skipDecl.

It would be a bit more generic with:

    else if (Token::simpleMatch(vartok, "decltype ("))
        vartok = vartok->linkAt(1)->next();

can you add some test like:

    for (decltype(x) *y : z)

There are AST tests in `test/testtokenizer.cpp`. There is for instance this test:

        ASSERT_EQUALS("forab:(", testAst("for (int a : b);"));

Somewhere below that I guess something like this would make sense:

        ASSERT_EQUALS("foryz:(", testAst("for (decltype(x) *y : z);"));

Reply:
Also your test should make the syntax tree correct ... but the types will probably not be determined properly.

If you compare the AST for :

     for (decltype(auto) x: items)

And:

    for (auto x: items)

I assume the type for `x` should be the same. To fix that I believe you can investigate `setValueTypeInTokenList`.